### ISSUE TITLE
Apply GP to Graph data

### ISSUE THREAD

**Issue by Apollo201** [cornellius-gp/gpytorch#1160]
Dear,

I would like to apply GP to graph data, and the following is my core code. Basic operation:
x=GCN(x)
res=GP1(x)

x=GCN(res.mean)
res=GP2(x)

Then I use **gpytorch.likelihoods.SoftmaxLikelihood.** The problem happens when I output the result, it looks without any learning and output randomly. Is there anything wrong with the code? thanks for your help ~

**Comment by jacobrgardner** [comment]
Hmm, I actually think the above should just about work out of the box. Can you share any training code, or ideally a small self contained example that “runs” but produces the problem?

**Comment by Apollo201** [comment]
Thanks for your reply, the following is core code, where GraphConv can be treated as the feature extractor. The problem is when I train the code, I found nothing can be learned by the model and it almost outputs random results (classification task with six categories).

![image](https://user-images.githubusercontent.com/19870456/83723902-f5963a80-a671-11ea-8852-5446b1ce531d.png)


```python
## The code depend on DGLGraph

import argparse, time
import numpy as np
import networkx as nx
import torch
import torch.nn as nn
import torch.nn.functional as F
from dgl import DGLGraph
from dgl.data import register_data_args, load_data
from dgl.nn.pytorch import GraphConv
import gpytorch
import math


class GaussianProcessLayer(gpytorch.models.ApproximateGP):
    def __init__(self, num_dim, grid_bounds=(-10., 10.), grid_size=32):
        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
            num_inducing_points=grid_size, batch_shape=torch.Size([num_dim])
        )
        variational_strategy = gpytorch.variational.MultitaskVariationalStrategy(
            gpytorch.variational.GridInterpolationVariationalStrategy(
                self, grid_size=grid_size, grid_bounds=[grid_bounds],
                variational_distribution=variational_distribution,
            ), num_tasks=num_dim,
        )
        super().__init__(variational_strategy)

        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(
                lengthscale_prior=gpytorch.priors.SmoothedBoxPrior(
                    math.exp(-1), math.exp(1), sigma=0.1, transform=torch.exp
                )
            )
        )
        self.mean_module = gpytorch.means.ConstantMean()
        self.grid_bounds = grid_bounds

    def forward(self, x):
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean, covar)

class DKLModel(torch.nn.Module):
    def __init__(self, g, in_feats, hidden_dim=16, out_dim=7, grid_bounds=(-10., 10.)):
        super(DKLModel, self).__init__()
        self.conv1 = GraphConv(in_feats, out_feats=hidden_dim)
        self.conv2 = GraphConv(in_feats=hidden_dim, out_feats=out_dim)
        self.out_dim= out_dim
        self.hidden_dim = hidden_dim
        self.gp_layer = GaussianProcessLayer(num_dim=out_dim, grid_bounds=grid_bounds)
        self.grid_bounds = grid_bounds
        self.g = g

    def forward(self, feature, mask):
        h = self.conv1(self.g, feature)
        x = self.conv2(self.g, h)[mask]
        features = gpytorch.utils.grid.scale_to_bounds(x, self.grid_bounds[0], self.grid_bounds[1])
        # This next line makes it so that we learn a GP for each feature
        features = x.transpose(-1, -2).unsqueeze(-1)
        res = self.gp_layer(features) # mean: 2708,16
        return res


def main(args):
    # load and preprocess dataset
    data = load_data(args)
    features = torch.FloatTensor(data.features)
    labels = torch.LongTensor(data.labels)

	train_mask = torch.BoolTensor(data.train_mask)
	val_mask = torch.BoolTensor(data.val_mask)
	test_mask = torch.BoolTensor(data.test_mask)

    in_feats = features.shape[1]
    n_classes = data.num_labels
    n_edges = data.graph.number_of_edges()

    g = DGLGraph(g)
    n_edges = g.number_of_edges()
    # normalization
    degs = g.in_degrees().float()
    norm = torch.pow(degs, -0.5)
    norm[torch.isinf(norm)] = 0

    model = DKLModel(g, in_feats, args.n_hidden, n_classes)

    likelihood = gpytorch.likelihoods.SoftmaxLikelihood(num_features=model.out_dim, num_classes=n_classes)
    mll = gpytorch.mlls.VariationalELBO(likelihood, model.gp_layer, num_data=len(torch.where(train_mask==True)))
    # initialize graph
    optimizer = torch.optim.SGD([
        {'params': model.conv1.parameters(), 'weight_decay': 1e-4},
        {'params': model.conv2.parameters(), 'weight_decay': 1e-4},
        {'params': model.gp_layer.hyperparameters(), 'lr': args.lr * 0.01},
        {'params': model.gp_layer.variational_parameters()},
        {'params': likelihood.parameters()},
    ], lr=args.lr, momentum=0.9, nesterov=True, weight_decay=0)

    dur = []
    for epoch in range(args.n_epochs):
        model.train()
        if epoch >= 3:
            t0 = time.time()
        # forward
        logits = model(features, torch.where(train_mask==True))
        loss = - mll(logits, labels[train_mask])
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
```

**Comment by Apollo201** [comment]
@gpleiss 
Hello, this is my basic code using GP to graph data where I use GCN as a feature extractor. I am quite confused that the output is almost random. Even though I pre-trained the feature extractor using standard supervised cross-entropy (can obtain good classification performance), then I use the pre-trained parameters to combine with GP, the accuracy still drop to random guess level. so did when I try to use different gp layer

```python
# GCN extractor
class Net(torch.nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = GCNConv(dataset.num_features, hidden_dim, cached=True)
        self.conv2 = GCNConv(hidden_dim, dataset.num_classes, cached=True)

    def forward(self):
        x = F.relu(self.conv1(data.X, data.A, data.W))
        x = self.conv2(data.X, data.A, data.W)
        return x

# This is code from gpytorch
class GaussianProcessLayer(gpytorch.models.ApproximateGP):
    def __init__(self, num_dim, grid_bounds=(-10., 10.), grid_size=64):
        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(
            num_inducing_points=grid_size, batch_shape=torch.Size([num_dim])
        )
        variational_strategy = gpytorch.variational.MultitaskVariationalStrategy(
            gpytorch.variational.GridInterpolationVariationalStrategy(
                self, grid_size=grid_size, grid_bounds=[grid_bounds],
                variational_distribution=variational_distribution,
            ), num_tasks=num_dim,
        )
        super().__init__(variational_strategy)

        self.covar_module = gpytorch.kernels.ScaleKernel(
            gpytorch.kernels.RBFKernel(
                lengthscale_prior=gpytorch.priors.SmoothedBoxPrior(
                    math.exp(-1), math.exp(1), sigma=0.1, transform=torch.exp
                )
            )
        )
        self.mean_module = gpytorch.means.ConstantMean()
        self.grid_bounds = grid_bounds

    def forward(self, x):
        mean = self.mean_module(x)
        covar = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean, covar)

# Incoporate GCN and GP
class DKLModel(gpytorch.Module):
    def __init__(self, num_dim, grid_bounds=(-10., 10.)):
        super(DKLModel, self).__init__()
        self.gp_layer = GaussianProcessLayer(inducing_points)
        self.grid_bounds = grid_bounds
        self.num_dim = num_dim

    def forward(self, gcn_extractor, mask=None):
        features = gcn_extractor()
        res = self.gp_layer(features)
        return res
		
model = DKLModel(num_dim = dataset.num_classes)
likelihood = gpytorch.likelihoods.SoftmaxLikelihood(num_features=model.num_dim, num_classes=dataset.num_classes
mll = gpytorch.mlls.VariationalELBO(likelihood, model.gp_layer, num_data=data.x)
gcn = Net()

def train_val():
    for i in range(100):
        model.train()
        likelihood.train()
        optimizer_gp.zero_grad()
        output = model(gcn, data.x)
        loss = -mll(output, data.y)
        loss.backward()
        optimizer_gp.step()

if __name__ == '__main__':
    train_val()
```


After that, I tried to use a different GP layer. The results is still terrible
```python
class StandardApproximateGP(gpytorch.models.ApproximateGP):
    def __init__(self, inducing_points):
        variational_distribution = gpytorch.variational.CholeskyVariationalDistribution(inducing_points.size(-2))
        variational_strategy = gpytorch.variational.VariationalStrategy(
            self, inducing_points, variational_distribution, learn_inducing_locations=True
        )
        super().__init__(variational_strategy)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
```
